packages/manager/modules/product-offers/src/product-offers.controller.js

```diff
@@ -0,0 +1,60 @@
+import OrderWorkflow from './workflows/product-offers-order-workflow.class';
+import ServicesWorkflow from './workflows/product-offers-services-workflow.class';
+
+import { WORKFLOW_TYPES } from './workflows/product-offers-workflow.constants';
+
+export default class ProductOffersController {
+  /* @ngInject */
+  constructor(
+    $q,
+    $translate,
+    ovhManagerProductOffersDetachService,
+    WucOrderCartService,
+  ) {
+    this.$q = $q;
+    this.$translate = $translate;
+    this.detachService = ovhManagerProductOffersDetachService;
+    this.WucOrderCartService = WucOrderCartService;
+  }
+
+  $onInit() {
+    if (!this.workflowType) {
+      throw new Error('ovhProductOffersComponent: workflowType is undefined');
+    }
+
+    if (!this.workflowOptions) {
+      throw new Error(
+        'ovhProductOffersComponent: workflowOptions is undefined',
+      );
+    }
+
+    switch (this.workflowType) {
+      case WORKFLOW_TYPES.ORDER:
+        this.workflow = new OrderWorkflow(
```

Comment:
It's weird that you have to pass angular dependencies manually to the constructor. I think it shows that OrderWokflow should be an angularjs service/factory/provider whatever because of the dependencies. And OrderWorkflow should then be injected in your controller

Reply:
Hmm I did this thinking in terms of inheritance, as Order/Services Workflow class instances. I will check to handle them through factory

Reply:
Ok in fact, I have an abstract class `Workflow`, which contains base methods for the two kind of workflows. `OrderWorkflow` and `ServiceWorkflow`, are children of `Workflow`. 
As all of them needs to use a service to make requests, that's why I'm injecting those angular dependencies.

Now I can create two factories for the two different workflows, but it means that I have to duplicate some code (especially for the pricing calculation), and I don't see too much benefits from.